core/market/decentralized/src/db/dao/agreement.rs

```diff
@@ -17,6 +20,37 @@ impl<'a> AsDao<'a> for AgreementDao<'a> {
 }
 
 impl<'c> AgreementDao<'c> {
+    pub async fn select(
+        &self,
+        id: &AgreementId,
+        validation_ts: NaiveDateTime,
+    ) -> DbResult<Option<Agreement>> {
+        let id = id.clone();
+        readonly_transaction(self.pool, move |conn| {
+            let mut agreement = match dsl::market_agreement
+                .filter(dsl::id.eq(&id))
+                .first::<Agreement>(conn)
+                .optional()?
+            {
+                None => return Ok(None),
+                Some(agreement) => agreement,
+            };
+
+            if agreement.valid_to < validation_ts {
+                agreement.state = AgreementState::Expired;
+                update_state(conn, &id, &agreement.state)?;
+            }
+
+            Ok(Some(agreement))
+        })
+        .await
+    }
+
+    pub async fn update_state(&self, id: &AgreementId, state: AgreementState) -> DbResult<bool> {
+        let id = id.clone();
+        do_with_transaction(self.pool, move |conn| update_state(conn, &id, &state)).await
+    }
+
```

Comment:
Important: We should check state of Agreement, before changing it, in the same transaction. Now we have dangerous pattern: we get Agreement in first transaction, check state and than change it in second transaction.
I don't like this either, that we are forced to implement something-like state machine on dao level :(

Reply:
added TODO